src/primaite/agents/aegis/gllm.py

```diff
@@ -101,10 +157,19 @@ def train_loop(model: GLLM, dataloader: DataLoader, network_desc: str, n_epochs:
             questions = batch["questions"]
             gllm_prompts = model.build_prompts(questions=questions, network_desc=network_desc, model="gllm")
 
-            gllm_texts, gllm_hidden_states = model(batch["graphs"], gllm_prompts)
-            gllm_responses.extend(gllm_texts)
-
-            loss = model.get_cosine_embeddings_loss(batch["gt_hidden_states"], gllm_hidden_states)
+            gllm_tokens, gllm_logprobs, gllm_hidden_states = model(batch["graphs"], gllm_prompts)
+            txt_response = model.llm.tokenizer.batch_decode(gllm_tokens, skip_special_tokens=False)[-1]
+            gllm_responses.append(txt_response)
+            if loss_fn == "cosine":
+                loss = model.get_cosine_embeddings_loss(batch["gt_hidden_states"], gllm_hidden_states)
+            elif loss_fn == "crossentropy":
+
+                loss = model.get_crossentropy_loss(
+                    gt_probs=batch["gt_logprobs"],
+                    gt_tokens=batch["gt_tokens"],
+                    gllm_probs=gllm_logprobs,
+                    gllm_tokens=gllm_tokens,
+                )
```

Comment:
does it make more sense to initialise this loss before the for loop, since it's global across the batches anyway

Reply:
dont htink it makes a difference, jsut a style thing

Reply:
nvm me